src/goto-instrument/contracts/contracts.cpp

```diff
@@ -914,6 +920,42 @@ void code_contractst::apply_loop_contract(
       throw 0;
     }
 
+    // By definition the `loop_content` is a set of instructions computed
+    // by `natural_loops` based on the CFG.
+    // Since we perform assigns clause instrumentation by sequentially
+    // traversing instructions from `loop_head` to `loop_end`,
+    // here we ensure that all instructions in `loop_content` belong within
+    // the [loop_head, loop_end] target range
+
+    // Check 1. (i \in loop_content) ==> loop_head <= i <= loop_end
+    for(const auto &i : loop_content)
+    {
+      if(
+        loop_head->location_number > i->location_number ||
+        i->location_number > loop_end->location_number)
```

Comment:
Could we add a regression test to cover this case somehow?

Reply:
I'll try to produce a test, discussion with @remi-delmas-3000 suggests that nested do-while loops are most likely to demonstrate the failing behaviour.

Reply:
I have added a test with nested loops, which demonstrates yet another type of problem (see comment in nested.desc), but I'll leave that for another PR.